eva/optimizer/rules/rules.py

```diff
@@ -499,6 +503,64 @@ def apply(self, before: LogicalJoin, context: OptimizerContext):
         yield new_join
 
 
+class ReorderPredicates(Rule):
+    """
+    The current implementation orders conjuncts based on their individual cost.
+    The optimization for OR clauses has `not` been implemented yet. Additionally, we do
+    not optimize predicates that are not user-defined functions since we assume that
+    they will likely be pushed to the underlying relational database, which will handle
+    the optimization process.
+    """
+
+    def __init__(self):
+        pattern = Pattern(OperatorType.LOGICALFILTER)
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        super().__init__(RuleType.REORDER_PREDICATES, pattern)
+
+    def promise(self):
+        return Promise.REORDER_PREDICATES
+
+    def check(self, before: LogicalFilter, context: OptimizerContext):
+        # there exists atleast one Function Expression
+        return len(list(before.predicate.find_all(FunctionExpression))) > 0
+
+    def apply(self, before: LogicalFilter, context: OptimizerContext):
+        # Decompose the expression tree into a list of conjuncts
+        conjuncts = to_conjunction_list(before.predicate)
+
+        # Segregate the conjuncts into simple and function expressions
+        contains_func_exprs = []
+        simple_exprs = []
+        for conjunct in conjuncts:
+            if list(conjunct.find_all(FunctionExpression)):
```

Comment:
Do we have a test case where there are no `FunctionExpression` in `conjunct` -- edge case?